Branch-SDK/src/main/java/io/branch/referral/ReferringUrlUtility.kt

```diff
@@ -0,0 +1,192 @@
+import org.json.JSONObject
+import java.text.SimpleDateFormat
+import java.util.*
+import androidx.annotation.VisibleForTesting
+import io.branch.referral.util.BranchEvent.ServerRequestLogEvent
+
+class ReferringUrlUtility (prefHelper: PrefHelper) {
+    private val urlQueryParameters: MutableMap<String, BranchUrlQueryParameter>
+    private var prefHelper: PrefHelper
+
+    init {
+        this.prefHelper = prefHelper
+        urlQueryParameters = deserializeFromJson(prefHelper.referringURLQueryParameters)
+        checkForAndMigrateOldGclid()
+    }
+
+    fun parseReferringURL(urlString: String) {
+        val uri = Uri.parse(urlString)
+
+        for (paramName in uri.queryParameterNames) {
+            val paramValue = uri.getQueryParameter(paramName)
+            PrefHelper.Debug("Found URL Query Parameter - Key: $paramName, Value: $paramValue")
+
+            if (isSupportedQueryParameter(paramName)) {
+                val param = findUrlQueryParam(paramName)
+                param.value = paramValue
+                param.timestamp = Date()
+                param.isDeepLink = true
+
+                // If there is no validity window, set to default.
+                if (param.validityWindow == 0L) {
+                    param.validityWindow = defaultValidityWindowForParam(paramName) as Long
+                }
+
+                urlQueryParameters[paramName] = param
+            }
+        }
+
+        prefHelper.setReferringUrlQueryParameters(serializeToJson(urlQueryParameters))
+        PrefHelper.Debug(prefHelper.referringURLQueryParameters.toString())
+    }
+
+    fun getURLQueryParamsForRequest(request: ServerRequest): JSONObject {
+        val returnedParams = mutableMapOf<String, Any>()
+
+        val gclid = addGclidValueFor(request)
+        if (gclid.length() > 0) {
+            val keys = gclid.keys()
+            while (keys.hasNext()) {
+                val key = keys.next()
+                returnedParams[key] = gclid.get(key)
+            }
+        }
+
+        return JSONObject(returnedParams as Map<*, *>)
+    }
+
+    @VisibleForTesting
+    private fun addGclidValueFor(request: ServerRequest): JSONObject {
+        val returnParams = JSONObject()
+
+        if (request is ServerRequestLogEvent || request is ServerRequestRegisterOpen) {
+            val gclid = urlQueryParameters[Defines.Jsonkey.Gclid.key]
+            if (gclid != null) {
+                if (gclid.value != null && gclid.value != PrefHelper.NO_STRING_VALUE) {
+                    returnParams.put(Defines.Jsonkey.Gclid.key, gclid.value)
+                    returnParams.put(Defines.Jsonkey.IsDeeplinkGclid, gclid.isDeepLink)
+
+                    gclid.isDeepLink = false
+                    prefHelper.setReferringUrlQueryParameters(serializeToJson(urlQueryParameters))
+                }
+            }
+        }
+
+        return returnParams
+    }
+
+    @VisibleForTesting
+    internal fun isSupportedQueryParameter(paramName: String): Boolean {
```

Comment:
Some functions are marked `internal`, another `private`, and some no modifier. No modifier in Kotlin is `public`, as opposed to Java which is `package-private`.

https://medium.com/@HugoMatilla/kotlin-basics-visibility-modifiers-public-internal-protected-and-private-c3bf972aee11

Regarding `@VisibleForTesting`: https://timkranen.medium.com/using-visiblefortesting-to-test-your-private-functions-c9e511fd1c2a

The tl;dr is (as i interpret it) be mindful of testing private functions since the public methods that use the results of them should be tested anyway and give you the same value (the outputs of your private functions). The tests are fine to keep but just something to keep in mind when thinking of how to write unit test.